crates/turborepo-filewatch/src/package_watcher.rs

```diff
@@ -0,0 +1,215 @@
+            tokio::select! {
+                _ = &mut self.exit_rx => return,
+                file_event = self.recv.recv().into_future() => match file_event {
+                    Ok(Ok(event)) => self.handle_file_event(event).await,
+                    // if we get an error, we need to re-discover the packages
+                    Ok(Err(_)) => self.rediscover_packages().await,
+                    Err(RecvError::Closed) => return,
+                    // if we end up lagging, just skip the event
+                    Err(RecvError::Lagged(count)) => {
+                        tracing::warn!("lagged behind {count} processing file watching events");
+                        continue;
+                    },
+                }
+            }
+        }
+    }
+
+    async fn handle_file_event(&mut self, file_event: Event) {
+        let glob_source_change = self.manager.workspace_glob_source(&self.repo_root);
+
+        tracing::trace!("file event: {:?}", file_event);
+
+        // if it is the package manager, update the glob list
+        let changed = if file_event
+            .paths
+            .iter()
+            .any(|p| glob_source_change.as_path().eq(p))
+        {
+            let new_filter = self
+                .manager
+                .get_workspace_globs(&self.repo_root)
+                // under some saving strategies a file can be totally empty for a moment
+                // during a save. these strategies emit multiple events and so we can
+                // a previous or subsequent event in the 'cluster' will still trigger
+                .unwrap_or_else(|_| self.filter.clone());
+
+            let changed = self.filter != new_filter;
+            self.filter = new_filter;
+            changed
+        } else {
+            false
+        };
+
+        if changed {
+            // if the glob list has changed, do a recursive walk and replace
+            self.rediscover_packages().await;
+        } else {
+            // if a path is not a valid utf8 string, it is not a valid path, so ignore
+            for path in file_event
+                .paths
+                .iter()
+                .filter_map(|p| p.as_os_str().to_str())
+            {
+                let path_file =
+                    AbsoluteSystemPathBuf::new(path).expect("watched paths are absolute");
+
+                // the path to the workspace this file is in is the parent
+                let path_workspace = path_file
+                    .parent()
+                    .expect("watched paths will not be at the root")
+                    .to_owned();
+
+                let is_workspace = match self
+                    .filter
+                    .target_is_workspace(&self.repo_root, &path_workspace)
+                {
+                    Ok(is_workspace) => is_workspace,
+                    Err(e) => {
+                        tracing::error!("error checking if path is in workspace: {:?}", e);
+                        continue;
```

Comment:
What can cause this error, and what are the implications of ignoring it? Are we going to have stale workspaces? Should we stop watching workspaces if we can't recover?

Reply:
Added a comment to explain and adjusted the log line.